doc/specs/#1595 - Suggestions UI/Snippets.md

```diff
@@ -0,0 +1,526 @@
+
+For the most part, we'll be using the [Suggestions UI] to display tasks to the
+user. This is a text cursor-relative UI surface that can quickly display actions
+to the user, in the context of what they're working on.
+
+The following are some examples from VsCode, Warp. These are meant to be
+illustrative of what these menus already look like in the wild:
+
+![](img/vscode-tasks-000.gif)
+
+![](img/warp-workflows-000.gif)
+
+A prototype of saving a command directly to the user's settings, then invoking
+it via the suggestions UI
+
+![](img/save-command.gif)
+
+A prototype of reading tasks from the CWD
+
+![](img/tasks-from-cwd.gif)
+
+
+## Tenets
+
+<table>
+
+<tr><td><strong>Compatibility</strong></td><td>
+
+I considered supporting YAML for local snippets (`.wt.json`), instead of JSON.
+JSON is not super friendly to command-lines - since everything's gotta be
+encapsulated as a string. Embedding tabs `\t`, newlines `\r`, escape characters,
+is fairly straightforward. However, quotes can get complicated fast in JSON,
+since they've got to be escaped too, and with many CLI utilities also having
+separate quote-parsing rules, JSON can get unwieldy quickly.
+
+However, supporting YAML directly would require us to spec out a YAML syntax for
+these files, and also find an OSS YAML parser and implement support for it. That
+would be quite a bit more expensive than JSON.
+
+</td></tr>
+
+<tr><td><strong>Accessibility</strong></td><td>
+
+Nothing particular to call out here. The Snippets pane will need to be a11y
+tested, like most of our other UI surfaces.
+
+</td></tr>
+
+<tr><td><strong>Sustainability</strong></td><td>
+
+No substantial climate impacts expected here. We're not using expensive compute
+resources for this feature, so the impact should be comparable to any other
+Terminal feature.
+
+</td></tr>
+
+<tr><td><strong>Localization</strong></td><td>
+
+I'm mildly worried here about the potential for community-driven tasks to have
+non-localized descriptions. We may need to accept a `description:{ en-us:"",
+pt-br:"", ...}`-style map of language->string descriptions. That may just need
+to be a future consideration for now.
+
+</td></tr>
+
+<tr><td><a name="security"></a><strong>Security</strong></td><td>
+
+Another reason we shouldn't support keys being able to be lazy-bound to local
+snippets: It's entirely too easy for `malicious.exe` to create a file in
+`%HomePath%` that creates a snippet for `\u003pwn-your-machine.exe\r` (or
+similar). Any app can read your settings file, and it is again too easy for that
+malicious app to set it's own action `id` to the same as some other well-meaning
+local snippet's ID which you DO have bound to a key.
+
+</td></tr>
+
+</table>
+
+### Other potential issues
+
+Something like `wt save ping 8.8.8.8 > foo.txt` isn't going to work the way
+users want. The shell is gonna get the first crack at parsing that commandline,
+and is going to try and redirect the _output of `wt`_ to `foo.txt`.

```

Comment:
Hmm... Are there any other potential commandline args/subcommands we'd be interested for snippets? Here's a few ideas to spitball:

### Idea 1
`wt save --output <save output to this file>`
`wt save --snippet <command to save>`

`--output` defaults to nothing
`--snippet` is optional

### Idea 2
`wt save output <path>`
`wt save snippet <command>`

This makes them mutually exclusive (essentially, the subcommand becomes `save-output` and `save-snippet`).

Reply:
I did drop notes in Future Considerations for this:

> * We may want to add additional params to the `save` subcommand in the future,
>   to configure where the snippet is saved:
>   * `--local`: Save to the `.wt.json` in the CWD, if there is one (or create
>     one)
>   * `--parent`: Save to the `.wt.json` in the first ancestor of the CWD, if
>     there is one. Otherwise create one here.
>   * `--settings`: Manually save to the settings file?
>   * `--profile`: save to this profile???? Not sure if this is actually possible.
>     Maybe with the `WT_SESSION_ID` env var to figure out which profile is in use
>     for the pane with that ID
>     * This would probably require p